Add EF Core setup, models, and repositories

AmmarAbdelateif0-0 opened the pull request:
This pull request introduces the initial Entity Framework Core setup for the project, including the database context, entity relationships, and the first migration to create the core tables. It also updates project dependencies and structure to support these changes.

**Entity Framework Core Integration and Database Modeling:**

- Added the `ApplicationDbContext` class in `Context/ApplicationDbContext.cs` to define the main database context, including `DbSet` properties for `User`, `Patient`, `Doctor`, and `Caregiver` entities, and configured their relationships (one-to-one, one-to-many) using Fluent API.
- Added the first migration (`Migrations/20251211221250_AddModels.cs` and `.Designer.cs`) to create the `Users`, `Doctors`, `Patients`, and `Caregivers` tables with appropriate columns, constraints, and indexes, reflecting the entity relationships defined in the context. [[1]](diffhunk://#diff-61d91427a6c707e88df209446118b550aebb90b8eedcf57d24e68e40a0185a58R1-R155) [[2]](diffhunk://#diff-34e3e1bfc5f888581774e7f0bc462ca0f602e243c47d05b7ea6ec0062665c345R1-R253)

**Project Structure and Dependency Updates:**

- Added the `Microsoft.EntityFrameworkCore.Proxies` package to enable features like lazy loading, and ensured all necessary EF Core packages are included in `BreastCancer.csproj`.
- Updated the project folder structure to include a new `Service` folder for future service layer implementations.

AmmarAbdelateif0-0 committed:
Add EF Core setup, models, and repositories

Configured `ApplicationDbContext` for database interactions and added a connection string in `appsettings.json`. Created entity models (`User`, `Doctor`, `Patient`, `Caregiver`) with relationships defined using Fluent API. Added migration `20251130112154_AddModels` to create database tables.

Implemented repository pattern with generic `Repository` and specific repositories (`UserRepository`, `DoctorRepository`, `PatientRepository`, `CaregiverRepository`) to abstract data access logic. Added interfaces for CRUD operations and entity-specific queries.

Updated `Program.cs` to register `ApplicationDbContext` and services. These changes establish a scalable architecture for the `BreastCancer` application.

mahmoud-40 commented on Context/ApplicationDbContext.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+
+        }
+        public DbSet<User> Users { get; set; }
```
Please consider using [lazy loading.](https://www.youtube.com/watch?v=QucztHO07ZI&t=182s)

AmmarAbdelateif0-0 replied:
Could you give me more details?

mahmoud-40 commented on Models/Doctor.cs:
```diff
@@ -0,0 +1,26 @@
+
+        [Range(0,60)]
+        public int YearsOfExperience{ get; set; }
+        public bool IsVerified{ get; set; }
```
isVerified should be by default false.

mahmoud-40 commented on Models/Caregiver.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class Caregiver
+    {
+        public int Id { get; set; }
```
The id should be String.

mahmoud-40 replied:
Apply for all other models

AmmarAbdelateif0-0 replied:
Will we use the Giud to generate Id?

mahmoud-40 commented on Models/Caregiver.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public int Id { get; set; }
+
+        public string RelationshipToPatient { get; set; }
```
try using better naming conviction like `RelationshipType `

mahmoud-40 commented on Models/Patient.cs:
```diff
@@ -0,0 +1,20 @@
+        public int Id { get; set; }
+
+        [MaxLength(2000)]
+        public string MedicalHistory { get; set; } 
```
I think this should be handled on a diff table.

AmmarAbdelateif0-0 replied:
Does the patient have many Medical History?

mahmoud-40 replied:
YESSSSSSS

mahmoud-40 commented on Models/User.cs:
```diff
@@ -0,0 +1,36 @@
+        public string Email { get; set; }
+
+        [DataType(DataType.Password)]
+        public string PasswordHash { get; set; }
```
We don't need to store passwords here as it's handled on keycloak side.

AmmarAbdelateif0-0 replied:
We really need to study KeyCloak

abdalahsalah replied:
We really need to study

mahmoud-40 replied:
I really need a wife :(

mahmoud-40 commented on Models/User.cs:
```diff
@@ -0,0 +1,36 @@
+
+        [DataType(DataType.Password)]
+        public string PasswordHash { get; set; }
+        public string Role { get; set; }
```
Why having a role field? we really don't need it as we have a table for each role.

mahmoud-40 commented on Models/User.cs:
```diff
@@ -0,0 +1,36 @@
+        [Phone]
+        public string Phone{ get; set; }
+
+        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
```
Try adding more audit fields for all models. 

- Created By - The user who created the record
- Created At - The date and time when the record was created
- Updated By - The user who last updated the record
- Updated At - The date and time when the record was last updated

AmmarAbdelateif0-0 committed:
Add repository services and update Swagger config

- Added `using` directives for repository interfaces and implementations.
- Registered `IPatientRepository`, `IUserRepository`, `IDoctorRepository`, and `ICaregiverRepository` in the DI container.
- Enabled repository pattern for better data access management.
- Updated Swagger configuration for "Breast Cancer API" (v1).

AmmarAbdelateif0-0 committed:
Refactor to generic repository and unit of work patterns

Replaced custom repositories with GenericRepository and IGenericRepository.
Introduced UnitOfWork and IUnitOfWork for centralized data access.
Changed entity primary keys from int to string and updated navigation
properties to support EF Core lazy loading. Added new fields to User
entity. Registered new services in DI and enabled lazy loading proxies.
Removed obsolete repository code and interfaces.

AmmarAbdelateif0-0 committed:
Refactor models to use string IDs and add soft deletion

- Change all primary keys and foreign keys to string (nvarchar)
- Add IsDeleted to User, Doctor, Patient, Caregiver for soft deletion
- Add Age, CreatedBy, UpdatedBy, UpdatedAt to User; remove PasswordHash and Role
- Replace Caregiver.RelationshipToPatient with RelationshipType
- Update migrations and model snapshot for new schema
- Enable lazy loading proxies in EF Core configuration

abdalahsalah reviewed (approved):
This is great
